internal/release/observed.go

```diff
@@ -0,0 +1,200 @@
+func IgnoreHookTestEvents(rel *ObservedRelease) {
+	if len(rel.Hooks) > 0 {
+		var hooks []helmrelease.Hook
+		for _, h := range rel.Hooks {
+			if !IsHookForEvent(&h, helmrelease.HookTest) {
+				hooks = append(hooks, h)
+			}
+		}
+		rel.Hooks = hooks
+	}
+}
+
+// ObservedRelease is a copy of a Helm release object, as observed to be written
+// to the storage by a storage.Observer. The object is detached from the Helm
+// storage object, and mutations to it do not change the underlying release
+// object.
+type ObservedRelease struct {
+	// Name of the release.
+	Name string `json:"name"`
+	// Version of the release, at times also called revision.
+	Version int `json:"version"`
+	// Info provides information about the release.
+	Info helmrelease.Info `json:"info"`
+	// ChartMetadata contains the current Chartfile data of the release.
+	ChartMetadata chart.Metadata `json:"chartMetadata"`
+	// Config is the set of extra Values added to the chart.
+	// These values override the default values inside the chart.
+	Config map[string]interface{} `json:"config"`
+	// Manifest is the string representation of the rendered template.
+	// It is omitted from the JSON representation of the object for performance
+	// reasons. ManifestChecksum can be used to notice changes, after which
+	// Manifest can be inspected while loaded.
+	Manifest string `json:"manifest"`
+	// Hooks are all the hooks declared for this release, and the current
+	// state they are in.
+	Hooks []helmrelease.Hook `json:"hooks"`
+	// Namespace is the Kubernetes namespace of the release.
+	Namespace string `json:"namespace"`
+	// Labels of the release.
+	Labels map[string]string `json:"labels"`
+}
+
+// Targets returns if the release matches the given name, namespace and
+// version. If the version is 0, it matches any version.
+func (o ObservedRelease) Targets(name, namespace string, version int) bool {
+	return o.Name == name && o.Namespace == namespace && (version == 0 || o.Version == version)
+}
+
+// Encode JSON encodes the ObservedRelease and writes it into the given writer.
+func (o ObservedRelease) Encode(w io.Writer) error {
+	enc := json.NewEncoder(w)
+	if err := enc.Encode(o); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ObserveRelease deep copies the values from the provided release.Release
+// into a new ObservedRelease while omitting all chart data except metadata.
+// If no filters are provided, it defaults to DefaultDataFilters. To not use
+// any filters, pass an explicit empty slice.
+func ObserveRelease(rel *helmrelease.Release, filter ...DataFilter) ObservedRelease {
+	if rel == nil {
+		return ObservedRelease{}
+	}
+
+	if filter == nil {
+		filter = DefaultDataFilters
+	}
+
+	obsRel := ObservedRelease{
+		Name:      rel.Name,
+		Version:   rel.Version,
+		Config:    nil,
+		Manifest:  rel.Manifest,
+		Hooks:     nil,
+		Namespace: rel.Namespace,
+		Labels:    nil,
+	}
+
+	if rel.Info != nil {
+		obsRel.Info = *rel.Info
```

Comment:
is `rel.Info` immutable? Why are we not copying it like the other fields?

Reply:
Because unlike all other structures, this structure is simple and can simply copy the values: https://github.com/helm/helm/blob/main/pkg/release/info.go#L23

Reply:
the filter option seems to be used only in the tests. Is there another use case? Otherwise maybe we can have a specific unexported func for filtering.

Reply:
It always uses the default filters, this might change depending on what actually looping through the results produces with current code. Which is why I made it part of the official API.